### PR TITLE
HOSTEDCP-1777, HOSTEDCP-1709: test/e2e: complete move to eventually blocks

### PR DESCRIPTION
**Pull request by stevekuznetsov**
test/e2e: complete move to eventually blocks

Signed-off-by: Steve Kuznetsov <skuznets@redhat.com>

---

~Depends on #4261~

### CODE REVIEW

**Comment by sjenning** [comment]
Seems to be failing immediate on DNS lookup failure waiting for Nodes.  How it gets by the guest apiserver check and then fails on the apiserver DNS lookup, I don't know :confounded: 
```
    util.go:1310: Waiting for guest client to become available
    util.go:137: waiting for kubeconfig to be published for HostedCluster e2e-clusters-llnmr/example-j46kj
    util.go:137: Successfully waited for kubeconfig to be published for HostedCluster e2e-clusters-llnmr/example-j46kj in 1m14.025s
    util.go:154: waiting for kubeconfig secret to have data
    util.go:154: Successfully waited for kubeconfig secret to have data in 25ms
    util.go:180: Waiting for a successful connection to the guest apiserver
    util.go:197: Successfully connected to the guest apiserver in 1m14s
    util.go:259: waiting for waiting for 3 nodes to become ready
    eventually.go:240: failed to get **v1.Node: failed to get API group resources: unable to retrieve the complete list of server APIs: v1: Get "https://api-example-j46kj.service.ci.hypershift.devcluster.openshift.com:443/api/v1": dial tcp: lookup api-example-j46kj.service.ci.hypershift.devcluster.openshift.com on 172.30.0.10:53: no such host
```

**Comment by stevekuznetsov** [comment]
May have used the wrong client, let me look ..

**Comment by stevekuznetsov** [comment]
```
    util.go:259: Successfully waited for waiting for 0 nodes to become ready in 31.025s
```

:thinking:

**Comment by stevekuznetsov** [comment]
```
=== NAME TestCreateClusterPrivate/ValidateHostedCluster
util.go:187: Failed to wait for a successful connection to the guest API server in 30m0s: context deadline exceeded
=== NAME TestCreateClusterPrivate
hypershift_framework.go:392: Summarizing unexpected conditions for HostedCluster example-4wjwc
hypershift_framework.go:405: ValidAWSIdentityProvider, Reason: InvalidIdentityProvider, Message: WebIdentityErr
hypershift_framework.go:405: ClusterVersionSucceeding, Reason: StatusUnknown, Message: Condition not found in the CVO.
hypershift_framework.go:405: ClusterVersionProgressing, Reason: StatusUnknown, Message: Condition not found in the CVO.
hypershift_framework.go:405: ClusterVersionReleaseAccepted, Reason: StatusUnknown, Message: Condition not found in the CVO.
hypershift_framework.go:396: Unknown condition ClusterVersionUpgradeable
hypershift_framework.go:405: ClusterVersionAvailable, Reason: StatusUnknown, Message: Condition not found in the CVO.
hypershift_framework.go:405: Degraded, Reason: UnavailableReplicas, Message: [capi-provider deployment has 1 unavailable replicas, kube-apiserver deployment has 1 unavailable replicas]
hypershift_framework.go:405: KubeAPIServerAvailable, Reason: WaitingForAvailable, Message: Waiting for Kube APIServer deployment to become available
hypershift_framework.go:405: Available, Reason: WaitingForAvailable, Message: Waiting for Kube APIServer deployment to become available
hypershift_framework.go:405: AWSEndpointAvailable, Reason: StatusUnknown, Message: AWSEndpointService conditions not found
hypershift_framework.go:405: AWSEndpointServiceAvailable, Reason: AWSError, Message: load balancer arn:aws:elasticloadbalancing:us-east-1:820196288204:loadbalancer/net/a8d48a79e53bf4edc8efeee1f8da6483/d3ba742c3fd7cfce is not yet active
hypershift_framework.go:405: IgnitionEndpointAvailable, Reason: NotFound, Message: Ignition server deployment not found
```

This has reproduced more than once now, unclear why ...

**Comment by stevekuznetsov** [comment]
We seem to be over budget on CPO mutations, but this PR does not change the CPO ...

```
    util.go:710: pod control-plane-operator-5b79bfc497-kzww9 over budget: budget: 5000, actual: 5264
```

I've seen this in some other unrelated PRs as well like the one to zip artifacts, so ignoring ...

**Comment by stevekuznetsov** [comment]
```
    eventually.go:220: observed *v1beta1.HostedCluster e2e-clusters-sgbmg/example-9rjkt invalid at RV 22194 after 10m0s: incorrect condition: wanted Degraded=False, got Degraded=True: UnavailableReplicas(certified-operators-catalog deployment has 1 unavailable replicas)
```

Seems like a flake.

/test e2e-azure

File: test/e2e/e2e_test.go [hunk 1]
```diff
@@ -246,7 +248,19 @@ func setupSharedOIDCProvider() error {
 	}
 	iamOptions.ParseAdditionalTags()
 
-	if _, err := iamOptions.CreateOIDCProvider(iamClient); err != nil {
+	createLogFile := filepath.Join(artifactDir, "create-oidc-provider.log")
```
Comment (enxebre) [inline test/e2e/e2e_test.go:251]: How would the flow look like for dev changing cmd code and running e2e locally?

File: hypershift-operator/controllers/hostedcluster/hostedcluster_controller.go [hunk 2]
```diff
@@ -1209,7 +1209,7 @@ func (r *HostedClusterReconciler) reconcile(ctx context.Context, req ctrl.Reques
 			}
 			return ctrl.Result{}, fmt.Errorf("failed to reconcile platform credentials: %w", err)
 		}
```
Comment (enxebre) [inline hypershift-operator/controllers/hostedcluster/hostedcluster_controller.go:1211]: Thanks for fixing this

**Comment by enxebre** [comment]
seems like we could have split this PR in two for the sake of better review? last three commits actually honor "complete move to eventually blocks", then we have logging changes and condition fixes? Are they connected somehow?

**Comment by enxebre** [comment]
needs rebase, lgtm otherwise

**Comment by stevekuznetsov** [comment]
@enxebre rebased, thanks

File: test/e2e/util/eventually.go [hunk 3]
```diff
@@ -84,22 +84,24 @@ func EventuallyObject[T client.Object](t *testing.T, ctx context.Context, object
 		option(opts)
 	}
 
-	t.Logf("waiting for %s", objective)
+	if os.Getenv("EVENTUALLY_VERBOSE") != "false" {
```
[the remark this answers is not in the document]
Comment (stevekuznetsov) [reply test/e2e/util/eventually.go:87]: No, where would you prefer it to be? My hope was it's not needed other than in the CI jobs, where it's already set

**Comment by stevekuznetsov** [comment]
> seems like we could have split this PR in two for the sake of better review? last three commits actually honor "complete move to eventually blocks", then we have logging changes and condition fixes? Are they connected somehow?

@enxebre yes - agreed totally. Our CI has been such a mess lately, though, that when I've done that each individual step has taken forever to merge so I was hoping not to have to deal with that :(

Logging is related in the sense that we're fixing e2e to be better.

Conditions are required since the `EventuallyCondition` construct as written can't implement the "expect this condition, if present, to have this status" semantic and I figured it would be better to make things explicit than to implement that.

File: hypershift-operator/controllers/hostedcluster/hostedcluster_controller.go (hunk 2 above)
Comment (stevekuznetsov) [reply hypershift-operator/controllers/hostedcluster/hostedcluster_controller.go:1211]: NP

File: test/e2e/e2e_test.go (hunk 1 above)
Comment (stevekuznetsov) [reply test/e2e/e2e_test.go:252]: You would need to set `$ARTIFACT_DIR`, then these would show up there. I agree that streaming in that case would be nice, I think we can add that in a follow-up by adding `os.Stdout` as a sink when `$EVENTUALLY_VERBOSE!="false"`.

**Comment by sjenning** [comment]
probably a rebase issue
```
# github.com/openshift/hypershift/test/e2e/util
test/e2e/util/util.go:203:36: assignment mismatch: 2 variables but WaitForGuestKubeConfig returns 1 value
make: *** [Makefile:185: e2e] Error 1
```

**Comment by stevekuznetsov** [comment]
@sjenning https://github.com/openshift/hypershift/pull/3849/files#diff-cebde6b83b4b372162b2c6a22710a4ab0b1838576120535593a66a3a9bdfc170 went in after my rebase and I needed another update. I will move that to use Eventually in a follow-up.

**Pull request merged**